### PR TITLE
KNEM: Adding initial support for KNEM SM

### PR DESCRIPTION
**Pull request by shamisp**
Missing functionality:
- check for capability flags
- async communication

### CODE REVIEW

**Comment by shamisp** [comment]
autobot build

**Comment by shamisp** [comment]
@alinask @yosefe please take a look

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com:8888/jenkins/job/gh-ucx-pr/749/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/186/ for details. (Mellanox internal link)

File: config/m4/knem.m4 [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+knem_happy="no"
+AC_ARG_WITH([knem],
+           [AS_HELP_STRING([--with-knem=(DIR)], [Enable the use of KNEM (default is NO).])],
+           [], [with_knem=no])
+
+AS_IF([test "x$with_knem" != xno],
+      [AS_IF([test ! -d $with_knem], 
+             [
+              AC_MSG_NOTICE([KNEM path was not found, guessing ...])
+              with_knem=$(${PKG_CONFIG} --variable=prefix knem || find /opt/knem* -name knem_io.h |xargs dirname |sed -e s,/include,,g)
```
Comment (miked-mellanox) [inline config/m4/knem.m4]: i think we should add --with-knem to be the ON by default, if no path specified - it will look in /opt and will disqualify itself if not found.

Comment (shamisp) [inline config/m4/knem.m4]: do we want me to update configure-devel only ? Or all of them ? What is used for jenkins ?

File: config/m4/knem.m4 [hunk 2]
```diff
@@ -0,0 +1,28 @@
+#
+# Check for KNEM support
+#
+knem_happy="no"
+AC_ARG_WITH([knem],
+           [AS_HELP_STRING([--with-knem=(DIR)], [Enable the use of KNEM (default is NO).])],
+           [], [with_knem=no])
```
Comment (miked-mellanox) [inline config/m4/knem.m4]: actually setting here: with_knem=guess will make it ON by default if available.

Comment (shamisp) [inline config/m4/knem.m4]: got it

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com:8888/jenkins/job/gh-ucx-pr/755/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/192/ for details. (Mellanox internal link)

**Comment by shamisp** [comment]
@miked-mellanox @alinask - Can you please share the failure ?

**Comment by yosefe** [comment]
It should be publicly accessible

**Comment by shamisp** [comment]
Not for me.

**Comment by shamisp** [comment]
I looked at the log and I don't understand. There are plenty valgrained complains in multiple places (not just this code). @miked-mellanox @yosefe @alinask  - need your help

**Comment by miked-mellanox** [comment]
@shamisp - can you connect to the public jenkins server 

* console http://bgate.mellanox.com:8888/jenkins/job/gh-ucx-pr/755/console
* valgrind http://bgate.mellanox.com:8888/jenkins/job/gh-ucx-pr/755/valgrindResult/

You can also have ssh access to the jenkins if you want to run same cmd line to repro.

**Comment by shamisp** [comment]
Can't access non standard ports. Can you please post the error. 10x

**Comment by alinask** [comment]
@shamisp the following tests failed when running with valgrind:

[  FAILED  ] test_ucp_wireup.one_sided_wireup, where TypeParam =  and GetParam() = 
[  FAILED  ] test_ucp_wireup.two_sided_wireup, where TypeParam =  and GetParam() = 
[  FAILED  ] test_ucp_tag.send_recv_exp, where TypeParam =  and GetParam() = 
[  FAILED  ] test_ucp_tag.send_recv_unexp, where TypeParam =  and GetParam() = 
[  FAILED  ] test_ucp_rma.blocking_put, where TypeParam =  and GetParam() = 
[  FAILED  ] test_ucp_rma.blocking_get, where TypeParam =  and GetParam() = 
[  FAILED  ] test_ucp_mmap.alloc, where TypeParam =  and GetParam() = 
[  FAILED  ] test_ucp_mmap.reg, where TypeParam =  and GetParam() = 
[  FAILED  ] test_ucp_atomic.atomic_add, where TypeParam =  and GetParam() = 
[  FAILED  ] test_ucp_atomic.atomic_fadd, where TypeParam =  and GetParam() = 
[  FAILED  ] test_ucp_atomic.atomic_swap, where TypeParam =  and GetParam() = 
[  FAILED  ] test_ucp_atomic.atomic_cswap, where TypeParam =  and GetParam() = 
[  FAILED  ] test_ucp_atomic.unaligned_atomic_add, where TypeParam =  and GetParam() = 
[  FAILED  ] knem/test_pd.reg/0, where GetParam() = knem/knem
[  FAILED  ] knem/uct_p2p_rma_test.put_zcopy/0, where GetParam() = knem/knem
[  FAILED  ] knem/uct_p2p_rma_test.put_zcopy/1, where GetParam() = knem/knem/loopback
[  FAILED  ] knem/uct_p2p_rma_test.get_zcopy/0, where GetParam() = knem/knem
[  FAILED  ] knem/uct_p2p_rma_test.get_zcopy/1, where GetParam() = knem/knem/loopback
[  FAILED  ] alloc_methods/test_mem.pd_alloc/0, where GetParam() = 1
[  FAILED  ] alloc_methods/test_mem.pd_alloc/1, where GetParam() = 2
[  FAILED  ] alloc_methods/test_mem.pd_alloc/2, where GetParam() = 3

**Comment by shamisp** [comment]
Can you please post and actual error from: 
[ FAILED ] knem/uct_p2p_rma_test.get_zcopy/0, where GetParam() = knem/knem
[ FAILED ] knem/uct_p2p_rma_test.get_zcopy/1, where GetParam() = knem/knem/loopback

**Comment by alinask** [comment]
[ RUN      ] knem/uct_p2p_rma_test.get_zcopy/0
[     INFO ] 1 3.2m 1x{1..3.2m} nocomp 2.5k 
/scrap/jenkins/jenkins/jobs/gh-ucx-pr/workspace/contrib/../src/ucs/gtest/test.cc:101: Failure
Failed
Got 8 valgrind errors during the test
[  FAILED  ] knem/uct_p2p_rma_test.get_zcopy/0, where GetParam() = knem/knem (314 ms)
[ RUN      ] knem/uct_p2p_rma_test.get_zcopy/1
[     INFO ] 1 3.2m 1x{1..3.2m} nocomp 2.5k 
/scrap/jenkins/jenkins/jobs/gh-ucx-pr/workspace/contrib/../src/ucs/gtest/test.cc:101: Failure
Failed
Got 8 valgrind errors during the test
[  FAILED  ] knem/uct_p2p_rma_test.get_zcopy/1, where GetParam() = knem/knem/loopback (308 ms)

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com:8888/jenkins/job/gh-ucx-pr/761/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/198/ for details. (Mellanox internal link)

**Comment by shamisp** [comment]
@openucx/maintainers please review

File: src/uct/sm/knem/knem_ep.c [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+UCS_CLASS_DEFINE_NEW_FUNC(uct_knem_ep_t, uct_ep_t, uct_iface_t*,
+                          const struct sockaddr *);
+UCS_CLASS_DEFINE_DELETE_FUNC(uct_knem_ep_t, uct_ep_t);
+
+
+#define uct_knem_trace_data(_remote_addr, _rkey, _fmt, ...) \
+     ucs_trace_data(_fmt " to %"PRIx64"(%+ld)", ## __VA_ARGS__, (_remote_addr), \
+                    (_rkey))
+
+#define UCT_KNEM_ZERO_LENGTH_POST(len)              \
```
Comment (yosefe) [inline src/uct/sm/knem/knem_ep.c]: indentation

File: src/uct/sm/knem/knem_iface.c [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+#include "knem_pd.h"
+#include "knem_iface.h"
+#include "knem_ep.h"
+
+#include <uct/api/addr.h>
+#include <uct/tl/context.h>
+
+UCT_PD_REGISTER_TL(&uct_knem_pd_component, &uct_knem_tl);
+
+static ucs_config_field_t uct_knem_iface_config_table[] = {
```
Comment (yosefe) [inline src/uct/sm/knem/knem_iface.c]: i don't think there is a need to define special config table for knem

File: src/uct/sm/knem/knem_iface.c [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    return UCS_OK;
+}
+
+static ucs_status_t uct_knem_iface_query(uct_iface_h tl_iface,
+                                       uct_iface_attr_t *iface_attr)
+{
+    memset(iface_attr, 0, sizeof(uct_iface_attr_t));
+
+    /* default values for all shared memory transports */
+    iface_attr->cap.put.max_zcopy      = 64*1024*1024; /* Pasha: tmp value */
```
Comment (yosefe) [inline src/uct/sm/knem/knem_iface.c]: can it be SIZE_MAX?

File: src/uct/sm/knem/knem_iface.c [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+static UCS_CLASS_DECLARE_DELETE_FUNC(uct_knem_iface_t, uct_iface_t);
+
+static uct_iface_ops_t uct_knem_iface_ops = {
+    .iface_close         = UCS_CLASS_DELETE_FUNC_NAME(uct_knem_iface_t),
+    .iface_query         = uct_knem_iface_query,
+    .iface_get_address   = uct_knem_iface_get_address,
+    .iface_is_reachable  = uct_knem_iface_is_reachable,
+    .iface_flush         = (void*)uct_knem_flush,
```
Comment (yosefe) [inline src/uct/sm/knem/knem_iface.c]: maybe use &ucs_empty_function_return_success here?

File: src/uct/sm/knem/knem_pd.c [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+
+    memset(&info, 0, sizeof(struct knem_cmd_info));
+
+    fd = open("/dev/knem", O_RDWR);
+    if (fd < 0) {
+        ucs_debug("Could not open the KNEM device file at /dev/knem: %m. Disabling knem resource");
+        *resources_p     = NULL;
+        *num_resources_p = 0;
+        return UCS_OK;
+    } else {
```
Comment (yosefe) [inline src/uct/sm/knem/knem_pd.c]: to make it more "stylish", can avoid the "else" clause, since there is already "return OK" in the "then" case, just continue the function code after the if

File: src/uct/sm/knem/knem_pd.c [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+            rc = ioctl(fd, KNEM_CMD_GET_INFO, &info);
+            if (rc < 0) {
+                *resources_p     = NULL;
+                *num_resources_p = 0;
+                close(fd);
+                ucs_debug("KNEM get info failed. not using knem, err = %d %m", rc);
+                return UCS_OK;
+            } else {
```
Comment (yosefe) [inline src/uct/sm/knem/knem_pd.c]: and same here

File: src/uct/sm/knem/knem_pd.c [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+
+static void uct_knem_pd_close(uct_pd_h pd)
+{
+    uct_knem_pd_t *knem_pd = (uct_knem_pd_t *)pd;
+    close(knem_pd->knem_fd);
+    ucs_free(knem_pd);
+}
+
+static ucs_status_t uct_knem_mem_reg(uct_pd_h pd, void *address, size_t length,
+                                   uct_mem_h *memh_p)
```
Comment (yosefe) [inline src/uct/sm/knem/knem_pd.c]: indentation

File: src/uct/sm/knem/knem_iface.c (hunk 5 above)
Comment (shamisp) [inline src/uct/sm/knem/knem_iface.c]: I can't test more in my VM. I would suggest to update it later.

Comment (yosefe) [inline src/uct/sm/knem/knem_iface.c]: what is the real limit imposed by VM here?  is it actually limiting knem copy size or something else?

Comment (shamisp) [inline src/uct/sm/knem/knem_iface.c]: VM == vmware - yes, the actual amount of available memory is very limited (1G), so max_int fails.

Comment (yosefe) [inline src/uct/sm/knem/knem_iface.c]: what fails exactly? knem copy? or memory allocation? if it is the latter, the correct place to fix is the tests

Comment (shamisp) [inline src/uct/sm/knem/knem_iface.c]: memory allocation - yes it is test issue. It tries to allocate the max defined by the capabilities and fails if it can't handle it

**Comment by shamisp** [comment]
@yosefe - Addressed the comments. Let's call it done :-)

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com:8888/jenkins/job/gh-ucx-pr/787/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com:8888/jenkins/job/gh-ucx-pr/788/ for details.

**Comment by shamisp** [comment]
@yosefe seems like MellanoxLab is unhappy, but we don't have an access to the log. I suspect that the error is not related to this patch.

File: src/uct/sm/knem/knem_iface.c (hunk 5 above)
Comment (yosefe) [inline src/uct/sm/knem/knem_iface.c]: so lets fix it in the test then. maybe add ucs_get_phys_mem_size() and use it to limit the test buffer size? see [this](http://nadeausoftware.com/articles/2012/09/c_c_tip_how_get_physical_memory_size_system)

Comment (shamisp) [inline src/uct/sm/knem/knem_iface.c]: Yes, probably we have to fix the test ? But, it should go as a separate PR

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/213/ for details. (Mellanox internal link)

File: src/uct/sm/knem/knem_iface.c (hunk 5 above)
Comment (yosefe) [inline src/uct/sm/knem/knem_iface.c]: see #347

Comment (shamisp) [inline src/uct/sm/knem/knem_iface.c]: Let's merge this one first. Then we will go back and update CMA and KNEM (both of them have this 64Mb restriction) as a part of #347

**Comment by yosefe** [comment]
+1
